ramips: add support for NetGear WN3100RPv2

rdesaintleger opened the pull request:
This device is the same as the WN3000RPv3.
Last versions of official firmware for this device are in fact WN3000RPv3 firmware.

Signed-off-by: Rodolphe de Saint Léger <rdesaintleger@gmail.com>

mans0n commented:
> Last versions of official firmware for this device are in fact WN3000RPv3 firmware.

Does that mean the stock firmware accepts both WN3000RPv3 and WN3100RPv2 images? I thought they are incompatible due to different `NETGEAR_HW_ID` and `NETGEAR_BOARD_ID`.

rdesaintleger commented:
No the two devices do not accept both images.

Il you look on the stock firmware sources which are downloadable from NetGEAR (for the WN3000RPv3), there is a second target which is defined in the makefile (WNRP3100RPv2). If you follow the build instructions, both targets are built.

I successfully build the stock firmware from the last sources of the WN3000RPv3 and flashed my device from the alternate image. I don't know why NetGear did not provide last versions for the WN3100RPv2.

mans0n commented:
Okay, thanks for the report.

I like the idea of reusing dts file by overriding `DEVICE_DTS`, but the normal approach is to create a separate DTS file. In this case, you can create a new DTSI file and move the common nodes to the DTSI. Please refer to the examples of [dtsi](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_cameo_810.dtsi) and [dts](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_trendnet_tew-810dr.dts).

And please add device specification and flashing instruction in the commit message. You may copy one from the WN3000RPv3 support commit.

mans0n commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -804,6 +804,24 @@ define Device/netgear_wn3000rp-v3
+  DEVICE_MODEL := WN3100RP
+  DEVICE_VARIANT := v2
+  DEVICE_DTS := mt7620a_netgear_wn3000rp-v3
+  SUPPORTED_DEVICES += wn3100rpv2 wn3000rpv3
```
Please remove this line. This line is not needed for newly added devices, and the images are not compatible anyway.

rdesaintleger replied:
I added it because I did encounter encounter a problem with sysupgrade.

Trying to upgrade the wn3100rpv2 image ended up in an error because the device now thinks it's a wn3000rpv3
Do you known where this information is computed/stored so the sysupgrade can work ?

mans0n replied:
Hmm, I guess that problem will disappear once you reinstall a (proper) new image.

rdesaintleger replied:
Ok, I'll come back once I've applied the requested changes

rdesaintleger commented:
> Okay, thanks for the report.
> 
> I like the idea of reusing dts file by overriding `DEVICE_DTS`, but the normal approach is to create a separate DTS file. In this case, you can create a new DTSI file and move the common nodes to the DTSI. Please refer to the examples of [dtsi](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_cameo_810.dtsi) and [dts](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_trendnet_tew-810dr.dts).
> 
> And please add device specification and flashing instruction in the commit message. You may copy one from the WN3000RPv3 support commit.

Ok, give me a few days for changing DTS recompile and test everything and I'll come back with new DTS.
I'll aslo add device specification in the commit. (I've found this information helpfull sor some devices)

Do I need to reopen another pull request for thise new changes ?

mans0n commented:
> Do I need to reopen another pull request for thise new changes ?

Nope, please stick to the original PR.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

mans0n commented:
Oh, I forgot to ask, have you tested all the features to make sure WN3000RPv3 and WN3100RPv2 have indeed the same hardware? I've experienced that a contributor say new device has the same hardware with existing device because it boots well, but it turns out to have different LED pins or something later.
I think that having both targets in the source is not enough to prove it, we need to compare the actual boards.

And could you provide the installation procedure?

rdesaintleger commented:
Hi,

Actuelly, I tested the snapshot image on one of my devices with success. It is currently running a backported version for the 21.02.2 (which I can also submit if you wish). I did not open the box, but I've made the following tests against the stable backport :
- Factory reset (sets OpenWRT with default 192.168.1.1 on the Ethernet port)
- LED green:router OK
- LED red: router OK,
- LED green: wifi OK,
- LED red: wifi OK,
- LED blue:leftarrow OK,
- LED blue:rightarrow OK,
- LED green:power OK,
- LED red:power OK,
- LED green:wps OK,
- Ethernet port OK,
- Wifi card as STA OK,
- Wifi card as AP (with single or multiple SSIDs) OK,
- Wifi card as Mesh node OK,
- Wifi card as AP+STA NOK, the card stops to receive broadcast packets on the STA side
- Wifi card as AP+Mesh NOK, the device do not connect to the Mesh.
- Wireguard VPN OK,
- opkg update/install OK

I believe that wireless combinations are due to missing support in driver (I've seen in forum that other people where having similar problems). This makes the devices unable to fullfill its original purpose as a Wifi repeater. The following tests has not been done:
- Opening the box the check for the UART port,
- VLAN support on Ethernet port (did test on previous images including a 19.07 one, but not snapshot)
 
 Flashing instructions are quite simple:
 - Use the factory image on the WebUI,
 - Upload the factory image using tftp recovery procedure
 - When using OpenWRT, upgrade using the sysupgrade image
 
 I'm currently updating the commit for the flash instructions. since most original features are working well, I'm pretty confident that hardware is the same for both devices and that missing features are due to missing support in current drivers/runtime. I would be happy to help for adding AP+STA, AP+Mesh or event AP+STA+Mesh (since the device supports it) but as last forum post regarding missing broadcast packets is 1 year old, I believe that no one is interested in using this kind of setup.
 
 Kind Regards.

mans0n commented:
Thanks for the detailed report. That was enough for me to believe that they have indeed the same hardware.

mans0n commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -804,6 +804,23 @@ define Device/netgear_wn3000rp-v3
+  DEVICE_VENDOR := NETGEAR
+  DEVICE_MODEL := WN3100RP
+  DEVICE_VARIANT := v2
+  SUPPORTED_DEVICES += wn3100rpv2
```
Please drop this line entirely unless it affects the factory image installation from the stock firmware.

rdesaintleger replied:
Sorry, but I'm confused about which line youre are talking about ? The whole block or a specific one VENDOR/MODEL/VARIANT or SUPPORTED_DEVICES ?

mans0n replied:
I mean only `SUPPORTED_DEVICES`.

rdesaintleger replied:
Ok, Ive applied changes (including dtsi renaming) on my local branch. I'll test recompilation, Web UI Flash and sysupgrade and I'll push an update once I've validated the new images.

mans0n commented:
I suggest to rename dtsi to something like `mt7620a_netgear_wn3x00rp.dtsi`. The current name is a bit long.

rdesaintleger commented:
Last tests results are the following:
- WebUI accept the factory image for flashing however the device does not reboot automatically,
- I've tested with (and without) SUPPORTED_DEVICES with the same behaviour (no reboot at the end of flash),
- Powercycling the device (after waiting a reasonable amount of time) reboots to OpenWRT
- TFTP procedure has no problem, the device flash and reboots to OpenWRT

I've updated flash instructions with this information. and detailed TFTP procedure

I've also tested sysupgrade which works without any problem once OpenWRT is booting.

rdesaintleger committed:
ramips: add support for Netgear WN3100RPv2

This patch adds support for the Netgear WN3100RPv2
http://www.netgear.com/support/product/wn3100rpv2.aspx

Specifications:
- SoC: MediaTek MT7620A (580MHz, ramips)
- RAM: 32MB DDR
- Storage: 8MB NOR SPI flash
- Wireless: builtin MT7620A, 2x2:2 with u.FL connectors
- Ethernet: 1x100M
- Stock firmware based on OpenWRT Kamikaze

Like the EX2700, the bootloader expects a secondary image signature,
see https://forum.openwrt.org/viewtopic.php?pid=312577#p312577

This device seems to be same hardware as a WN3000RPv3

Flash instructions:
- Use the Netgear WebUI to upgrade to OpenWRT using the factory image (see note below),
- Use the sysupgrade image for upgrading versions from OpenWRT,
- TFTP recovery procedure can be used to flash the factory image (preferred method).

Note:
- The WebUI may not reboot automatically, wait at least 5 minutes before powercycling the device

Flashing using TFTP:
- Set you IP address to 192.168.1.10/24 (no gateway)
- Connect your machine to the Ethernet port
- Power off the device and wait for 10 seconds,
- Hold the reset button and power on the device (do not release reset),
- Hold the reset button until the green light is flashing (Approx. 15s)
- launch tftp, set mode to binary and connect to 192.168.1.1
- put the factory firmware image
- All leds will switch off (like a power off), this is normal
- Wait for the device to reboot in the new OpenWRT image (max 10 mins)
- The first boot will take longer than usual.
- After boot, the Device IP on the thernet port is 192.168.1.1

Signed-off-by: Rodolphe de Saint Léger <rdesaintleger@gmail.com>

rdesaintleger commented:
Sorry for the multiple pushes, I made mistakes in commit description (wrong title). There is no changes in the commited files

mans0n commented:
Oops, sorry. I already merged your last patch patch as 7bc20cb6143e...

mans0n commented:
Reverted and re-added as 46c5de53856d.
